[Work-in-Progress] Download a certificate of completion(#76)

Deep-Chill opened the pull request:
Allows downloading a certificate of completion at the /sequence_slug/certificate URL after all elements in the sequence have been viewed.
- `settings.py`: Updated logger to only show warning messages and not the font subsetting and optimization messages that show up due to fontTools

Deep-Chill committed:
Download a certificate of completion

Deep-Chill committed:
minor changes

smirolo commented on pages/urls/views/elements.py:
```diff
@@ -24,9 +24,12 @@
+from ...views.elements import PageElementView, CertificateDownloadView
 
 urlpatterns = [
+    re_path(r'^(?P<sequence_slug>%s)/certificate' % settings.SLUG_RE,
```
Add a '/' to the URL. By convention, API endpoints: no trailing '/', Pages/Downloads: trailing '/'.

smirolo commented on pages/views/elements.py:
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
         return context
+
+
+class CertificateDownloadView(LoginRequiredMixin, DetailView):
```
I am not sure we want to have a `LoginRequiredMixin` here. The djaodjin-pages app is used within other projects that handle role-based access control (RBAC) through URL decorators.

smirolo commented on pages/views/elements.py:
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
+
+        has_completed_sequence = True
+        for element in enumerated_elements:
+            user_element_progress = user_enumerated_progress.filter(rank=element.rank).first()
```
Relies on N (maximum number of elements) queries to the database for comparing two sequences.

smirolo commented on pages/views/elements.py:
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
+                break
+
+        if has_completed_sequence:
+            last_element = enumerated_elements.last().page_element
```
Based on the code above, I guess we rely on the last (certificate) element to have `min_viewing_duration == 0`.

smirolo commented on pages/views/elements.py:
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
+            context.update({
+                'user': self.request.user,
+                'sequence': sequence,
+                'current_date': date.today(),
```
Interesting. We rely on the current Web server date.
1. It could be important to pass the timezone as a query parameter just to make sure a one off day does not trigger e-mail and calls to tech support.
2. The date should most likely be fixed so that two successive downloads do not return different date. This might be important for CEUs and other certificate that expire, triggering taking the course again.

smirolo reviewed (commented):
I will have to write tests first to check various corner cases before merging the PR. Thank you!

Deep-Chill committed:
Code refactor

- Updated URL
- Removed LoginRequiredMixin
- Added completion date to SequenceProgress to record date of completion
- Improved database query to check completion of all items
- Added check for Certificate items and ignoring them in the View

Current issues: Still uses the server time.

Deep-Chill committed:
Update elements.py

smirolo commented on pages/views/elements.py:
```diff
@@ -246,19 +246,36 @@ def get_context_data(self, **kwargs):
+            # Exclude last element from the check, if it is a certificate
+            enumerated_elements = enumerated_elements.exclude(pk=enumerated_elements.last().pk)
+
+        completed_elements_subquery = user_enumerated_progress.filter(
```
Really cool subquery using `OuterRef`. Does that change the minimum Django version required to run djaodjin-pages?

smirolo reviewed (approved):
Looks really good. Thank you.
